### PR TITLE
Claim arkeo when claiming from ETH

### PR DESCRIPTION
**Pull request by 0xean**
PR is somewhat untested with waiting to determine the best way to get integration tests built out but it should not affect any arkeo module functionality and continues to build out needed functionality in the claim module.

Summary of changes:

1. Adds keeper dependencies to claimKeeper for the bankKeeper and AccountKeeper to help with transferring tokens during claiming process
2. Modifies claim records to include a single coin vs coins for rewards and all associated functionality.  
3. Adds a new param for the module to determine the amount of dust all airdrop users should get prior to claiming
4. Moves `MakeTestCodec` into a util file so it can be shared across modules and avoid circular imports
5. fixes genesis export for claim module and needed params storage mechanism 
6. Adds `ClaimCoinsForAction` that will be called to transfer coins to users for various actions and update their records (need integ tests)

### CODE REVIEW

File: app/app.go
Comment (johnnyluo) [inline app/app.go:514]: I think if `claimmodulekeeper.NewKeeper` don't return a pointer , then here we don't need to de-pointer it

File: x/claim/genesis.go
Comment (johnnyluo) [inline x/claim/genesis.go:15]: I think when `InitGenesis` failed , it is ok to panic , because the blockchain should not start anyway

File: x/claim/genesis.go
Comment (johnnyluo) [inline x/claim/genesis.go:27]: Also here , I think if export record fail, it should just panic

File: x/claim/keeper/claim.go
Comment (johnnyluo) [inline x/claim/keeper/claim.go:69]:
```suggestion
	var claimRecords []types.ClaimRecord
```
My IDE tells me to use this instead

File: x/claim/keeper/claim.go
Comment (johnnyluo) [inline x/claim/keeper/claim.go:135]: I think probably it is a good idea to check  the following
```
if len(claimRecord.ActionCompleted) != 3 {
   return sdk.Coin{}, nil
}
```
If somehow `claimRecord.ActionCompleted` is empty , then , this line will panic

File: x/claim/keeper/claim.go
Comment (johnnyluo) [inline x/claim/keeper/claim.go:213]: if it failed to save claimrecord , I think we should `return sdk.Coin{},err` , otherwise people can claim it again , right?

File: x/claim/keeper/claim.go
Comment (johnnyluo) [inline x/claim/keeper/claim.go:235]: here , comment should be consistent with function like
```
// CreateModuleAccount xxxxx
```

File: x/claim/keeper/claim_test.go
Comment (johnnyluo) [inline x/claim/keeper/claim_test.go:32]: Here , if we can add a few more cases , like when `ActionCompleted` is empty , or when `ActionCompleted` only 1 / 2 items in it.  make sure it still pass the test, and the function doesn't cause panic

File: x/claim/keeper/claim_test.go
Comment (johnnyluo) [inline x/claim/keeper/claim_test.go:169]: we could write a function like `GetRandomArkeoAddr()` , and then reuse it everywhere

File: app/app.go
Comment (0xean) [reply app/app.go:514]: This is the default from scaffolding, in general I am bit confused why we aren't more generally passing around pointers in the codebase, so am curious why your preference would be to pass this by value?

File: x/claim/keeper/claim.go
Comment (0xean) [reply x/claim/keeper/claim.go:69]: any specific reason? are they not functionally identical?

File: x/claim/keeper/claim.go
Comment (0xean) [reply x/claim/keeper/claim.go:213]: returning the claimableAmount here doesn't enable them to claim again as the err isn't nil, but yea, I see no reason to return the claimableAmount in that state.

File: x/claim/keeper/claim_test.go
Comment (0xean) [reply x/claim/keeper/claim_test.go:32]: yup, this is a good test. will add.

File: x/claim/keeper/msg_server_claim_eth.go
Comment (0xean) [reply x/claim/keeper/msg_server_claim_eth.go:65]: good catch.

File: x/claim/keeper/claim_test.go
Comment (0xean) [reply x/claim/keeper/claim_test.go:32]: I added a guard to ensure that we cannot set a record without the correct number of items in the array, I then added tests to ensure that it works correctly to avoid setting them.  Let me know if you think this is sufficient.

Comment (0xean) [reply x/claim/keeper/claim_test.go:32]: https://github.com/arkeonetwork/arkeo/pull/41/files#diff-7103aae3d6697d5e85af8e61d198a523f535578d293edbc33aaa373e401bba08R166

File: app/app.go
Comment (johnnyluo) [reply app/app.go:514]: Yup, I understand it is default from scaffolding , it all depends on how we use it , here we de-pointer it, so instead we could just return by value

File: x/claim/keeper/msg_server_claim_eth.go
Comment (johnnyluo) [inline x/claim/keeper/msg_server_claim_eth.go:38]: Based on the document , there will have claim record setup at genesis , I think those claim record will be useful for the first time user trying to claim. After first claim , we already create an additional claim record for `arkeo` chain,
So given that , instead of update the status as `ActionCompleted` , can we just delete it?  because we don't need it anymore .

The first 1/3 of token has been sent to their arkeo address, and the rest 2/3 will be credit to their arkeo address when they stake / vote.

File: x/claim/keeper/claim.go
Comment (johnnyluo) [reply x/claim/keeper/claim.go:73]: your way is totally legit as well , I guess it is just preference , I am using `Goland` it always recommend me to do it this way, I will leave it to you to decide.

File: x/claim/keeper/msg_server_claim_eth.go
Comment (johnnyluo) [inline x/claim/keeper/msg_server_claim_eth.go:60]: This make me to think we should flat out `ActionCompleted` to more concrete fields  like
```
ForeignChainClaim bool
Stake bool
Vote bool
```
By using `ActionComplete` as a slice , it is not easy to understand , I need to always trying to figure out what the first `true` means , second `false` means. etc.

File: x/claim/keeper/msg_server_claim_eth.go
Comment (johnnyluo) [inline x/claim/keeper/msg_server_claim_eth.go:63]: emit event should be moved to the end , since at this stage , no fund has been credit to user's account , if could error out after this step , then we might end up a scenario user didn't get fund , but event is on chain

File: x/claim/keeper/msg_server_claim_eth.go
Comment (johnnyluo) [inline x/claim/keeper/msg_server_claim_eth.go:77]: I think [the](https://github.com/arkeonetwork/arkeo/pull/41/files#diff-2e39f49e8b029fe4d783197ca9d2cf7f4abcf97a9ec1188a76e2d820e2a57dbdR60) , it should be `false`
However , in `ClaimCoinsForAction` we should double check whether the action has been claimed or not

File: app/app.go
Comment (0xean) [reply app/app.go:514]: okay, will revise.

File: x/claim/keeper/msg_server_claim_eth.go
Comment (0xean) [reply x/claim/keeper/msg_server_claim_eth.go:38]: > So given that , instead of update the status as ActionCompleted , can we just delete it? because we don't need it anymore .

We could, i decided to leave the record so that if a user cam back to query it, they would see that they already claimed vs showing that there was nothing ever to claim.  Happy to delete if storage concerns on the node, etc make it worth removing the record

File: x/claim/keeper/claim.go
Comment (0xean) [reply x/claim/keeper/claim.go:243]: what about the initial mint?

File: x/claim/keeper/msg_server_claim_eth.go
Comment (0xean) [reply x/claim/keeper/msg_server_claim_eth.go:77]: it is false now and we do check in ClaimCoinsForAction

File: x/claim/keeper/msg_server_claim_eth.go
Comment (johnnyluo) [reply x/claim/keeper/msg_server_claim_eth.go:63]: nope , for example , if line 71 fail , then the event will be write to block , however the user didn't get anything

File: x/claim/keeper/claim.go
Comment (johnnyluo) [reply x/claim/keeper/claim.go:243]: Generally we don't create module account manually, as module account is not created based on seed, the address is just hash of the module name , and if you want to mint , we write the logic into `genesis.go` file. for example
```
// Mint coins into the reserve
	if data.Reserve > 0 {
		coin := common.NewCoin(common.RuneNative, cosmos.NewUint(data.Reserve))
		if err := keeper.MintToModule(ctx, ModuleName, coin); err != nil {
			panic(err)
		}
		if err := keeper.SendFromModuleToModule(ctx, ModuleName, ReserveName, common.NewCoins(coin)); err != nil {
			panic(err)
		}
	}
```

File: x/claim/keeper/msg_server_claim_eth.go
Comment (johnnyluo) [reply x/claim/keeper/msg_server_claim_eth.go]: Cool , we could do that in a seperate PR , don't want to hold off this one for any longer

File: x/claim/keeper/msg_server_claim_eth.go
Comment (0xean) [reply x/claim/keeper/msg_server_claim_eth.go:63]: > PostHandlers: [PostHandler](https://docs.cosmos.network/main/core/baseapp#posthandler)s run after the execution of the message. If they fail, the state change of runMsgs, as well of PostHandlers, are both reverted.

This seems contradictory to the documentation.  If a transaction errors, and the state is not reverted, we have other logical problems that need to be dealt with and this behavior would be quite odd.

Comment (0xean) [reply x/claim/keeper/msg_server_claim_eth.go:63]: I will create a ticket for further investigation into this, because its problematic if these transaction are not atomic as you suggest.

Additionally, we have a different issue here as well where a user could clobber an existing claim record if trying to assign to an arkeo address that already has a drop to claim. 

Will create issues for both of these and we can address in the future while keeping this moving forward.

Comment (0xean) [reply x/claim/keeper/msg_server_claim_eth.go:63]: https://github.com/arkeonetwork/arkeo/issues/43
https://github.com/arkeonetwork/arkeo/issues/42

Comment (johnnyluo) [reply x/claim/keeper/msg_server_claim_eth.go:63]: Looking at the source code again, I think you are right , if we return error , then the state will not be changed.

**Review by johnnyluo — APPROVED** [review]
LGTM

**Pull request merged**
